### PR TITLE
[Launcher]: Support StopFramework

### PR DESCRIPTION
**Pull request by yqwang-ms**

### CODE REVIEW

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java [hunk 1]
```diff
@@ -287,6 +288,56 @@ private void upgradeFramework(FrameworkRequest frameworkRequest) throws Exceptio
...
+    }
+
+    updateExecutionType(frameworkRequest);
+  }
+
+  private void updateExecutionType(FrameworkRequest frameworkRequest) throws Exception {
+    String frameworkName = frameworkRequest.getFrameworkName();
+    ExecutionType executionType = frameworkRequest.getFrameworkDescriptor().getExecutionType();
+
+    if (executionType == ExecutionType.STOP) {
```
Comment (fanyangCS) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java]: >STOP [](start = 39, length = 4)

STOP is the only executionType handled in this function. do you plan to extend the executionType later on?

Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java]: maybe extend it.

---
In reply to: [171839789](https://github.com/Microsoft/pai/pull/346#discussion_r171839789) [](ancestors = 171839789)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java [hunk 2]
```diff
@@ -888,13 +897,13 @@ public void onFrameworkRequestsUpdated(Map<String, FrameworkRequest> frameworkRe
   }
 
   // Cleanup Framework level external resource [HDFS, RM] before RemoveFramework
```
Comment (sterowang) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java]: Add "." after the sentence?

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java [hunk 3]
```diff
@@ -594,34 +601,36 @@ private void launchApplication(FrameworkStatus frameworkStatus, ApplicationSubmi
     statusManager.transitionFrameworkState(frameworkName, FrameworkState.APPLICATION_LAUNCHED);
   }
 
-  private void createApplication(FrameworkStatus frameworkStatus) throws Exception {
+  private void createApplication(FrameworkStatus frameworkStatus, boolean usedToStop) throws Exception {
```
Comment (sterowang) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java]: >usedToStop [](start = 74, length = 10)

"stopping" would be a more appropriate name I think.

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/RequestManager.java [hunk 4]
```diff
@@ -173,6 +175,53 @@ private void deleteOrphanFrameworks() throws Exception {
...
+  }
+
+  private void stopFrameworkRequestInternal(String frameworkName) throws Exception {
+    updateExecutionTypeInternal(frameworkName, ExecutionType.STOP);
+  }
+
+  // stopOrphanFrameworks need to be handled in WebServer side instead of AM side,
+  // since AM is not always running, such as when the FrameworkState is not APPLICATION_RUNNING.
+  private void stopOrphanFrameworks() throws Exception {
+    // A Framework is StopOrphan, if and only if its ParentFramework is not null and Stopped.
```
Comment (sterowang) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/RequestManager.java]: Why “stopOrphanFrameworks" do not exists before?

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java [hunk 5]
```diff
@@ -482,12 +533,15 @@ public synchronized void transitionFrameworkState(
     frameworkStatus.setFrameworkState(dstState);
 
     // Update ZK Status
-    zkStore.setFrameworkStatus(frameworkName, frameworkStatus);
-    LOGGER.logInfo("Transitioned Framework [%s] from [%s] to [%s]", frameworkName, srcState, dstState);
+    if (!event.getSkipToPersist()) {
+      zkStore.setFrameworkStatus(frameworkName, frameworkStatus);
```
Comment (sterowang) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java]: >zkStore.setFrameworkStatus(frameworkName, frameworkStatus); [](start = 6, length = 59)

When will the stopped framework to be persisted?

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java (hunk 3 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java]: but it is better to use verb than noun after to.
And eventually we stop it :)

---
In reply to: [172148127](https://github.com/Microsoft/pai/pull/346#discussion_r172148127) [](ancestors = 172148127)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/RequestManager.java (hunk 4 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/RequestManager.java]: Previously, we only have delete interface, the counterpart is deleteOrphanFrameworks

---
In reply to: [172149640](https://github.com/Microsoft/pai/pull/346#discussion_r172149640) [](ancestors = 172149640)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java (hunk 5 above)
Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/StatusManager.java]: At line 338

---
In reply to: [172149878](https://github.com/Microsoft/pai/pull/346#discussion_r172149878) [](ancestors = 172149878)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java (hunk 3 above)
Comment (sterowang) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/service/Service.java]: "isPlaceholderApplication" would be better.

---
In reply to: [172154514](https://github.com/Microsoft/pai/pull/346#discussion_r172154514) [](ancestors = 172154514,172148127)

**Pull request merged**